### ISSUE TITLE
[MNG-7026] Ability to only build direct dependants

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8010]
**[Chadi](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chadiem)** opened **[MNG-7026](https://issues.apache.org/jira/browse/MNG-7026?redirect=false)** and commented

We are an organization with a large reactor. When we update a project X, we would like to have the possibility to compile the artifacts that directly depend on X, without having to recompile everything.

The fix itself is trivial and consists of traversing in non-recursive mode. I can prepare a PR if you are OK with the request.

Its activation could be done using a flag such as --amdd, --also-make-direct-dependents, or maybe --also-make-dependents=direct

What do you think?

---

**Affects:** 3.6.3

1 votes, 2 watchers

**Comment by jira-importer** [comment]
**[Martin Kanters](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=martinkanters)** commented

I think this indeed would be relatively easy to fix, but I am not sure whether this is something we would like to support as it seems to be quite an exotic case. 

Perhaps it helps to say that you can exclude certain projects with the `--projects !` flag if it is only certain submodules that you want to avoid to build?

**Comment by jira-importer** [comment]
**[Chadi](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chadiem)** commented

Thanks Martin for the quick feedback.

Perhaps it might not be needed with smaller projects, but in our case, we have thousands of projects and having this would be highly beneficial.

Using exclusions as you're suggesting, or explicitly listing the projects in a multi-module pom is an issue because those have to be explicitly maintained and continuously updated.

Basically X is a client project. I need to make changes to X, and by using -amd alone, it recompiles the whole tree of dependents. But my dependencies are clean, so by having something like -amd+direct, I make sure that I only compile projects that use X, and no more. This is very helpful for development purposes, or even in a (give-me-a-quick-feedback) CI context where I'm looking for a build result without having to build the whole reactor.

**Comment by jira-importer** [comment]
**[Martin Kanters](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=martinkanters)** commented

Thanks for coming back so quickly. I do not work with such big projects, so I do not share your experience, but I do understand what you mean now.

One thing is on my mind. In the case that you have these projects: C depends on B depends on A. You mean that if you change A, you only want to rebuild B and not C using something like -amd. 
While it's less common, it's possible that C could fail by the changes A introduces. Is that something that you don't mind in your use-case?

**Comment by jira-importer** [comment]
**[Chadi](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chadiem)** commented

That's correct, and this is something we do not mind indeed.

We've actually spent a significant amount of time cleaning up the dependencies using the maven-dependency-plugin (+failOnWarning), so code dependencies that come from transitive dependencies are rare in our case.

**Comment by jira-importer** [comment]
**[Martin Kanters](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=martinkanters)** commented

Right, then I guess it would make sense for you indeed. I'm a bit hesitant in building this in, because it (still) seems like a niche feature. It could definitely work out for you, but it will introduce extra complexity in the codebase and also in the usage/understandability of the flags. I'm wondering whether something like this can be built into an extension, but I do not have sufficient experience in that area.
I hope some other experienced Maven users/contributors can weigh in.
